src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,211 @@
+import linear_algebra.finite_dimensional
+import measure_theory.group.fundamental_domain
+
+/-!
+# ℤ-lattices
+
+Let `E` be a finite dimensional vector space over a `normed_lattice_field` `K` that is also
+a `floor_ring`, e.g. `ℚ` or `ℝ`. A (full) ℤ-lattice `L` of `E` is a discrete subgroup of `E` such
+that `L` spans `E` over `K`.
+
+The ℤ-lattice `L` can be defined in two ways:
+* For `b` a basis of `E`, then `L : submodule.span ℤ (set.range b)` is a ℤ-lattice of `E`.
+* As `L : add_subgroup E` with the additional properties:
+  `(hd : ∀ r : ℝ, (L ∩ (metric.closed_ball 0 r)).finite)`, that is `L` is discrete
+  `(hs : submodule.span ℝ (L : set E) = ⊤)`, that is `L` spans `E` over `K`.
+
+## Main result
+* `zspan.is_add_fundamental_domain`: for `L : submodule.span ℤ (set.range b)` a ℤ-lattice, proves
+ that the set defined by `zspan.fundamental_domain` is indeed a fundamental domain.
+
+-/
+
+open_locale big_operators classical
+
+noncomputable theory
+
+namespace zspan
+
+open measure_theory measurable_set submodule
+
+variables {E ι : Type*}
+
+section normed_lattice_field
+
+variables {K : Type*} [normed_lattice_field K]
+variables [normed_add_comm_group E] [normed_space K E]
+variables (b : basis ι K E)
+
+/-- The fundamental domain of the ℤ-lattice spanned by `b`. See `zspan.is_add_fundamental_domain`
+for the proof that it is the fundamental domain. -/
+def fundamental_domain : set E := { m | ∀ i, b.repr m i ∈ set.Ico (0 : K) 1 }
+
+@[simp]
+lemma mem_fundamental_domain {m : E} :
+  m ∈ fundamental_domain b ↔ ∀ i, b.repr m i ∈ set.Ico (0 : K) 1 := iff.rfl
+
+variables [floor_ring K]
+
+section fintype
+
+variable [fintype ι]
+
+/-- The map that sends a vector of `E` to the element of the ℤ-lattice spanned by `b` obtained
+by rounding down its coordinates on the basis `b`. -/
+def floor (m : E) : span ℤ (set.range b) := ∑ i, ⌊b.repr m i⌋ • b.restrict_scalars ℤ i
+
+@[simp]
+lemma repr_floor_apply (m : E) (i : ι) :
+  b.repr (floor b m) i = ⌊b.repr m i⌋ :=
+by simp only [floor, zsmul_eq_smul_cast K, b.repr.map_smul, finsupp.single_apply,
+  finset.sum_apply', basis.repr_self, finsupp.smul_single', mul_one, finset.sum_ite_eq',
+  finset.mem_univ, if_true, coe_sum, coe_smul_of_tower, basis.restrict_scalars_apply,
+  linear_equiv.map_sum]
+
+/-- The map that sends a vector `E` to the fundamental domain of the lattice,
+see `zspan.fract_mem_fundamental_domain`. -/
+def fract : E → E := λ m, m - floor b m
+
+lemma fract_apply (m : E) : fract b m = m - floor b m := rfl
+
+@[simp]
+lemma repr_fract_apply (m : E) (i : ι):
+  b.repr (fract b m) i = int.fract (b.repr m i) :=
+by rw [fract, map_sub, finsupp.coe_sub, pi.sub_apply, repr_floor_apply, int.fract]
+
+@[simp]
+lemma fract_fract (m : E) : fract b (fract b m) = fract b m :=
+basis.ext_elem b (λ _, by simp only [repr_fract_apply, int.fract_fract])
+
+@[simp]
+lemma fract_zspan_add (m : E) {v : E} (h : v ∈ span ℤ (set.range b)) :
+  fract b (v + m) = fract b m :=
```

Comment:
Can we also have
```
lemma fract_add_zspan (m : E) {v : E} (h : v ∈ span ℤ (set.range b)) :
  fract b (m + v) = fract b m :=
```
Also, a better name might be
```suggestion
lemma fract_add_of_left_mem_span (m : E) {v : E} (h : v ∈ span ℤ (set.range b)) :
  fract b (v + m) = fract b m :=
```

Reply:
I prefer to keep the names `fract_zspan_add` and `fract_add_zspan` to mimic the names `int.fract_int_add` and `int.fract_add_int`.